### ISSUE TITLE
Design: Configs per session, and removing duplicated functions in a Component

### ISSUE THREAD

**Issue by aliabid94** [gradio-app/gradio#4774]
Two issues to be resolved with the same approach:

1) Right now, a Gradio Blocks app has a single config. We need there to be one config *per user session*, so that we can support:
  a. Backend parameter updates (e.g. type, Dropdown choices)
  b. Different components for different sessions (needed for 4.0 rendering)
2) For a component, the functions `__init__`, `get_config` and `update` have essentially the same parameters. This leads to lots of code duplication, and will be annoying for custom components in 4.0.

Proposal:
1. Instead of storing component properties as attributes of the object, we store them in a new Config object

CHANGE:
```diff
class Textbox(...):
    def __init__(lines=..., placeholder=....)
-        self.lines = lines
-        self.placeholder = placeholder
+        cfg = Config()
+        cfg.lines = lines
+        cfg.placeholder = placeholder
```
2. All Component backend methods take a Config, and use that for logic. This Config is stored per user session, similar to States.

CHANGES:
```diff
class Number(...):
-    def preprocess(self, x: float | None) -> float | None:
-        if self.minimum is not None and x < self.minimum:
+    def preprocess(self, x: float | None, cfg: Config) -> float | None:
+        if cfg.minimum is not None and x < cfg.minimum:
            ...
```

3. Remove the `get_config` function entirely. Wherever we need it, we simply get the `Config` object, whose initial state is stored in the Component object, and whose updated states are tracked per user session.

4. Remove `Component.update` entirely. Instead the user returns a Component object with the values to change. To do this, we will have to:
  a. Set all `__init__` default values to None, and overwrite None with actual default values in the constructor body. This allows us to know when a user has overriden values when the Component is being used as an update
  b. Use thread locals to know if a Component is initialized inside an event listener, and to treat it as an update if so. 

CHANGES:
```diff
def update_dropdown():
-    return gr.Dropdown.update(choices=["A", "B", "C", value="A"])
+    return gr.Dropdown(choices=["A", "B", "C", value="A"])
```

**Comment by abidlabs** [comment]
I really like it. Just some quick thoughts:

1. Could we make this backwards-compatible and merge it into `main`? Two big advantages to that: (1) this will affect the backend code a lot (in a good way), which will be good to have in `main` to avoid merge conflicts that happen because I'm cleaning up the backend code for all of the components (2) The more code we can merge into `main` the better, so that we can test it out thoroughly before 4.0 launch

I think the only thing we need to do in order to make it backwards-compatible is to make `gr.Dropdown.update()` an alias for `gr.Dropdown()` and so on for each component

2. It *might* be a little confusing for people who are reading code to realize that returning a component from a function only updates the new parameters, it doesn't create a new component from scratch. Tbh I don't think it's a big deal, but we should maybe keep it in mind

**Comment by freddyaboulton** [comment]
Nice proposal @aliabid94 !

Rather than a general `Config` that's shared across all component classes (no type hints) what if we just type the argument as an instance of the component itself? For example:

```python
def preprocess(self, x: float | None, state: gr.Number)
```
That should enable type hints and I think it's clear what it means and it corresponds to the "stateful" model we're working towards.

I'm don't follow this:
>  Set all __init__ default values to None, and overwrite None with actual default values in the constructor body. This allows us to know when a user has overriden values when the Component is being used as an update

Won't the `None`s be converted back to the defaults since the `__init__` is still running? 

I like the idea of getting rid of `update`. I think the best way to do that would be if all event handlers got the state of the component in that session. That way everything is an update and users can just do 

```python
dropdown.choices = [...]
dropdown.value = ...
return dropdown
```
which is something we have seen users ask about. However, that is not backwards compatible. Still thinking about this point. The other thing we can do is have `update` be auto-generated somehow in a way that preserves the type hints. Something similar to how we did event handlers. That way we can achieve less code duplication.

Also, where are the configs being stored? In the backend? If so, I think we really need a way to clear session data after disconnecting otherwise we're going to OOM eventually (especially viral apps).

PS: today I learned about the diff markdown in github
